### DISCUSSION TITLE
psycopg3 how to COPY data from one table to another

### DISCUSSION THREAD

**Discussion by samuelecat** [psycopg/psycopg#285] [Q&A] [answered]
Googling I could not find any example about how to copy data from one table on a remote server to another table sitting on a different remote server.
I've also read the doc https://www.psycopg.org/psycopg3/docs/basic/copy.html and there was nothing like this so I'm still confused.

here is how I'm currently doing it:
```python
src_pool = ConnectionPool(conninfo=make_conninfo(**src_db_credentials))
dst_pool = ConnectionPool(conninfo=make_conninfo(**dst_db_credentials))

with src_pool.connection() as src_conn:
    with dst_pool.connection() as dst_conn:
        with src_conn.cursor() as src_cursor:
            with src_cursor.copy("COPY (SELECT col1,col2 FROM src_table WHERE col1=123) TO STDOUT") as src_copy:
                with dst_conn.cursor() as dst_cursor:
                    dst_cursor.execute("CREATE TABLE IF NOT EXISTS dst_table(col1 INT PRIMARY KEY, col2 TEXT)")
                    with dst_cursor.copy("COPY dst_table (col1,col2) FROM STDIN") as dst_copy:
                        for row in src_copy.rows(): # this way ???
                            dst_copy.write_row(row) # this way ???
```

what is the proper way without losing data or impact resources and performance?
thank you

**Comment by dvarrazzo** [comment]
Hello,

This way of doing should work, but you are converting every field to a Python object and then back to postgres, so I assume you are doing some unneeded CPU work on the client.

If you read and write data block by block you can skip in the Python conversion: you can use:

```python
for data in src_copy:
    dst_copy.write(data)
```

If source and destination table have exactly the same data types (not only compatible types, such as int4-> int8) you can try to use binary copy (using `FORMAT BINARY` in the COPY statements) which should be lighter on the CPU of the two servers.

It would be interesting if you run a comparison of the three methods and give us some timings. We could add the example to the docs. Cheers!

**Comment by samuelecat** [comment]
Hello, 
thank you for your fast response.

Replacing these two working lines of code:
```python
for row in src_copy.rows():
    dst_copy.write_row(row)
```

with:
```python
for data in src_copy:
    dst_copy.write(data)
```

Produces this error:
```
Traceback (most recent call last):
  File "/test/.venv/lib/python3.8/site-packages/psycopg/cursor.py", line 671, in copy
    yield copy
  File "copy_01.py", line 21, in <module>
    dst_copy.write(data)
  File "/test/.venv/lib/python3.8/site-packages/psycopg/copy.py", line 262, in write
    data = self.formatter.write(buffer)
  File "/test/.venv/lib/python3.8/site-packages/psycopg/copy.py", line 485, in write
    data = self._ensure_bytes(buffer)
  File "/test/.venv/lib/python3.8/site-packages/psycopg/copy.py", line 513, in _ensure_bytes
    raise TypeError(f"can't write {type(data).__name__}")
TypeError: can't write memoryview

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "copy_01.py", line 21, in <module>
    dst_copy.write(data)
  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/test/.venv/lib/python3.8/site-packages/psycopg/cursor.py", line 671, in copy
    yield copy
  File "/test/.venv/lib/python3.8/site-packages/psycopg/copy.py", line 211, in __exit__
    self.finish(exc_val)
  File "/test/.venv/lib/python3.8/site-packages/psycopg/copy.py", line 279, in finish
    self.connection.wait(self._end_copy_in_gen(exc))
  File "/test/.venv/lib/python3.8/site-packages/psycopg/connection.py", line 770, in wait
    return waiting.wait(gen, self.pgconn.socket, timeout=timeout)
  File "/test/.venv/lib/python3.8/site-packages/psycopg/waiting.py", line 237, in wait_epoll
    s = gen.send(ready)
  File "/test/.venv/lib/python3.8/site-packages/psycopg/copy.py", line 163, in _end_copy_in_gen
    res = yield from copy_end(self._pgconn, bmsg)
  File "/test/.venv/lib/python3.8/site-packages/psycopg/generators.py", line 230, in copy_end
    raise e.error_from_result(result, encoding=encoding)
psycopg.errors.QueryCanceled: COPY from stdin failed: error from Python: TypeError - can't write memoryview
CONTEXT:  COPY dst_table, line 1
```

my `requirements.txt` contains, among other stuff, the following lines:
```
psycopg[binary,pool]==3.0.11
psycopg-binary==3.0.11
psycopg-pool==3.1.1
```
running Ubuntu Linux 20.04 x86_64

**Reply by dvarrazzo** [reply]
Oh, I think this is caused by #254. I think we may consider this a bug and release the fix in 3.0.12 rather in 3.1.

In the meantime you could try:

```python
for data in src_copy:
    dst_copy.write(bytes(data))
```

**Comment by samuelecat** [comment]
with
```python
for data in src_copy:
    dst_copy.write(bytes(data))
```
works again, thank you.

I've also tried to test a copy 5 times with and 5 times without `WITH (FORMAT BINARY)` the latter seems marginally faster (22 seconds instead of 25) however the network conditions aren't optimal and I can't exclude that the test servers aren't running anything else.

**Reply by dvarrazzo** [reply]
Cool, thank you.

Is `copy.write()` noticeably faster than `copy.write_row()`?

**Reply by samuelecat** [reply]
I need to test it in isolation and properly monitor cpu and memory usage but I would say, yes

**Reply by dvarrazzo** [reply]
I have released psycopg 3.0.12, which includes a fix for #254, so the explicit `bytes()` call is no more necessary.

I have also added [an example in the docs](https://www.psycopg.org/psycopg3/docs/basic/copy.html#example-copying-a-table-across-servers), showing how to copy data across servers.

I will add an answer to provide the full example cleanup.

**Comment by dvarrazzo** [comment, marked as answer]
Using some cleanup to your code, your sample could be written as:

```python
src_pool = ConnectionPool(kwargs=src_db_credentials)
dst_pool = ConnectionPool(kwargs=dst_db_credentials)

with src_pool.connection() as src_conn, dst_pool.connection() as dst_conn:
    dst_conn.execute("CREATE TABLE IF NOT EXISTS dst_table(col1 INT PRIMARY KEY, col2 TEXT)")
    with src_conn.cursor().copy(
        "COPY (SELECT col1,col2 FROM src_table WHERE col1=123) TO STDOUT (FORMAT BINARY)"
    ) as src_copy:
        with dst_conn.cursor().copy("COPY dst_table (col1,col2) FROM STDIN (FORMAT BINARY)") as dst_copy:
            for data in src_copy:
                dst_copy.write(data)
```

- Using `write()` instead of `write_row()`, there will be no conversion of the data to and from Python objects.
- `FORMAT BINARY` is more efficient than the implicit default `FORMAT TEXT`, but it can only be used if the tables have identical schema; [see the docs](https://www.psycopg.org/psycopg3/docs/basic/copy.html#example-copying-a-table-across-servers) for details.
- We can skip on the cursors context blocks to avoid to drift too much on the right. Client-side cursors don't take resources to require an explicit finalization apart from normal GC.
- Unrelated to the question, but the connection pool can take a `kwargs` argument to avoid the use of `make_conninfo()`.
